Mono.Cecil.Cil/MethodBody.cs

```diff
@@ -306,126 +306,193 @@ void RemoveSequencePoint (Instruction instruction)
-			// Local scopes created during editing will also likely be resolved (so no numerical offsets).
-			// So while local scopes which are unresolved are relatively rare if they appear, manipulating them based
-			// on the offsets allone is pretty hard (since we can't rely on correct offsets of instructions).
-			// On the other hand resolved local scopes are easy to maintain, since they point to instructions and thus inserting
+			// Manipulating unresolved InstructionOffsets is pretty hard (since we can't rely on correct offsets of instructions).
+			// On the other hand resolved InstructionOffsets are easy to maintain, since they point to instructions and thus inserting
 			// instructions is basically a no-op and removing instructions is as easy as changing the pointer.
 			// For this reason the algorithm here is:
 			//  - First make sure that all instruction offsets are resolved - if not - resolve them
-			//     - First time this will be relatively expensinve as it will walk the entire method body to convert offsets to instruction pointers
-			//       Almost all local scopes are stored in the "right" order (sequentially per start offsets), so the code uses a simple one-item
-			//       cache instruction<->offset to avoid walking instructions multiple times (that would only happen for scopes which are out of order).
-			//     - Subsequent calls should be cheap as it will only walk all local scopes without doing anything
-			//     - If there was an edit on local scope which makes some of them unresolved, the cost is proportional
+			//     - First time this will be relatively expensive as it will walk the entire method body to convert offsets to instruction pointers
+			//       Within the same debug info, IL offsets are typically stored in the "right" order (sequentially per start offsets),
+			//       so the code uses a simple one-item cache instruction<->offset to avoid walking instructions multiple times
+			//       (that would only happen for scopes which are out of order).
+			//     - Subsequent calls should be cheap as it will only walk all local scopes without doing anything (as it checks that they're resolved)
+			//     - If there was an edit which adds some unresolved, the cost is proportional (the code will only resolve those)
 			//  - Then update as necessary by manipulaitng instruction references alone
 
-			InstructionOffsetCache cache = new InstructionOffsetCache () {
-				Offset = 0,
-				Index = 0,
-				Instruction = items [0]
-			};
+			InstructionOffsetResolver resolver = new InstructionOffsetResolver (items, removedInstruction, existingInstruction);
 
-			UpdateLocalScope (debug_info.Scope, removedInstruction, existingInstruction, ref cache);
+			if (method.debug_info != null)
+				UpdateLocalScope (method.debug_info.Scope, ref resolver);
+
+			var customDebugInfo = method.custom_infos ?? method.debug_info?.custom_infos;
+			if (customDebugInfo != null) {
+				foreach (var custom_debug_info in customDebugInfo) {
```

Comment:
Am I correct in that this is basically the only interesting change -- also walk the state machine and async method infos, instead of just the local scope?

Reply:
Yes - that's the gist of it.
The rest is mainly refactoring to reuse the same code which was used only for the local scopes.